### PR TITLE
FIX-#1930: Fix one of the cases of heterogeneous data for read_csv

### LINKED ISSUE #1930: Different from pandas index at `read_csv` in case of specified `names`
### System information
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Any
- **Modin version** (`modin.__version__`): 0.8.0
- **Python version**: 3.7.5
- **Code we can use to reproduce**:
```python
if __name__ == "__main__":
    import pandas
    import modin.pandas as pd
    from modin.pandas.test.test_io import (
        create_test_pandas_dataframe,
        TEST_CSV_FILENAME,
        df_equals,
    )

    create_test_pandas_dataframe().to_csv(TEST_CSV_FILENAME)

    md_df = pd.read_csv(TEST_CSV_FILENAME, names=["c1", "c2", "c3", "c4", "c5"])
    pd_df = pandas.read_csv(TEST_CSV_FILENAME, names=["c1", "c2", "c3", "c4", "c5"])

    try:
        df_equals(md_df, pd_df)
    except Exception as e:
        print(e)  # DataFrame.index are different

    md_df = pd.read_csv(
        TEST_CSV_FILENAME, names=["c1", "c2", "c3", "c4", "c5"], usecols=[0, 1, 2, 3, 4]
    )
    pd_df = pandas.read_csv(
        TEST_CSV_FILENAME, names=["c1", "c2", "c3", "c4", "c5"], usecols=[0, 1, 2, 3, 4]
    )

    df_equals(md_df, pd_df)  # everything is okay

```

### Describe the problem
Here is the output of the code above:
```
DataFrame.index values are different (100.0 %)
[left]:  RangeIndex(start=0, stop=5, step=1)
[right]: Float64Index([nan, 0.0, 1.0, 2.0, 3.0], dtype='float64')
```
As we can see the indices are different. In case `names` parameter is specified `read_csv` will start read data right from the first line of the file and that line pandas indicates with `nan` value at index, when Modin just creates `RangeIndex`. But when `usecols` is specified at the same time as `names` Pandas creates `RangeIndex` the same as Modin.

### PR DESCRIPTION
**Pull request by anmyachev**
Signed-off-by: Anatoly Myachev <anatoly.myachev@intel.com>

## What do these changes do?

- [x] first commit message and PR title follow format outlined [here](https://modin.readthedocs.io/en/latest/development/contributing.html#commit-message-formatting)
  > **_NOTE:_**  If you edit the PR title to match this format, you need to add another commit (even if it's empty) or amend your last commit for the CI job that checks the PR title to pick up the new PR title.
- [x] passes `flake8 modin/ asv_bench/benchmarks scripts/doc_checker.py`
- [x] passes `black --check modin/ asv_bench/benchmarks scripts/doc_checker.py`
- [x] signed commit with `git commit -s` 
- [x] Resolves #1930 
- [x] tests added and passing
- [x] module layout described at `docs/development/architecture.rst` is up-to-date

### CODE REVIEW

**Comment by anmyachev** [comment]
@YarShev I found a interesting problem with Unidist. Could you take a look? 
Job link: https://github.com/modin-project/modin/actions/runs/3781849458/jobs/6535203451

```bash
    unwrapped_kwargs = {k: unwrap_object_refs(v) for k, v in kwargs.items()}
/usr/share/miniconda3/envs/modin_on_unidist/lib/python3.8/site-packages/unidist/core/backends/common/utils.py:26: in unwrap_object_refs
    unwrapped_value = unwrap_object_refs(
/usr/share/miniconda3/envs/modin_on_unidist/lib/python3.8/site-packages/unidist/core/backends/common/utils.py:26: in unwrap_object_refs
    unwrapped_value = unwrap_object_refs(
E   RecursionError: maximum recursion depth exceeded while calling a Python object
!!! Recursion detected (same locals & position)

---------- coverage: platform linux, python 3.8.15-final-0 -----------
Coverage XML written to file coverage.xml

=========================== short test summary info ============================
FAILED modin/pandas/test/dataframe/test_indexing.py::test_loc_multi_index - RecursionError: maximum recursion depth exceeded while calling a Python object
```

**Comment by YarShev** [comment]
@Retribution98 is looking into this.

**Comment by anmyachev** [comment]
> @anmyachev, we merged a fix in unidist master. Can you change requirements/env_unidist.yml to install latest unidist master to see if the issue is gone?

the issue is fixed. 
Do I need to skip the test before the new release?

**Comment by YarShev** [comment]
@anmyachev, we released unidist 0.2.2, which contains the fix. The release is already available on conda-forge so you should just rerun tests and they should pass with the new release.

File: requirements/env_unidist.yml [hunk 1]
```diff
@@ -52,3 +51,4 @@ dependencies:
       - connectorx>=0.2.6a4
       # The `numpydoc` version should match the version installed in the `lint-pydocstyle` job of the CI.
       - numpydoc==1.1.0
+      - git+https://github.com/modin-project/unidist.git@master#egg=unidist[mpi]
```
Comment (YarShev) [inline requirements/env_unidist.yml:54]: Let's revert the change. Tests should pass with the new release.

**Comment by anmyachev** [comment]
> @anmyachev did you make sure your newly added test is failing without your fix?

yes, I rechecked
